File: field_force/field_force/page/requisition_report/requisition_report.py
import frappe
import json
import datetime
from frappe.utils import pretty_date, now, add_to_date
from field_force.field_force.page.utils import generate_excel_and_download, get_time_in_12_hour_format

@frappe.whitelist()
def execute(filters=None):
    columns = get_columns()
    data = get_absolute_data(filters)
    return data, columns

def get_absolute_data(filters, export=False):
    query_result = get_query_data(filters)

    for requisition in query_result:
        requisition['action'] = get_appropiate_action_button(requisition)
        requisition['status'] = get_workflow_state_with_color(requisition)
        requisition['docname'] = requisition['name']
        requisition['date'] = frappe.format(requisition.transaction_date, 'Date')
        requisition['delivery_date'] = frappe.format(requisition.delivery_date, 'Date')
        requisition['expected_delivery_date'] = frappe.format(requisition.expected_delivery_date, 'Date')
        requisition['pretty_date'] = custom_pretty_date(requisition.modified)
        requisition['difference'] = requisition.get("total_qty") - requisition.get("total_accepted_qty")

        if not export:
            set_link_to_doc(requisition, 'name', 'requisition', color='#006433')
            set_link_to_doc(requisition, 'customer', 'customer', color='dodgerblue')
            set_link_to_doc(requisition, 'territory', 'territory', color='teal')
            set_link_to_doc(requisition, 'department', 'department', color='slateblue')
            requisition['print'] = get_pdf_button("Requisition", requisition['docname'])

        else:
            requisition['status'] = requisition.workflow_state

    return query_result

def get_conditions(filters):
    from_date = filters.get('from_date')
    to_date = filters.get('to_date')
    customer = filters.get('customer')
    status = filters.get('status')
    territory = filters.get('territory')
    department = filters.get('department')
    partner_group = filters.get('partner_group')
    delivery_trip_created = filters.get('delivery_trip_created')

    conditions = {}

    if from_date and to_date:
        conditions["transaction_date"] = ["between", [from_date, to_date]]
    if customer:
        conditions["customer"] = customer
    if status:
        conditions['workflow_state'] = status
    if territory:
        conditions['territory'] = territory
    if department:
        conditions['department'] = department
    if partner_group:
        conditions['partner_group'] = partner_group
    if delivery_trip_created:
        conditions['delivery_trip_created'] = delivery_trip_created
    # else:
    #     conditions['delivery_trip_created'] = 0

    if "Warehouse User" in frappe.get_roles(frappe.session.user) and not delivery_trip_created:
        conditions['delivery_trip_created'] = 0

    # return " and ".join(conditions)
    return conditions

def get_query_data(filters):
    try:
        filters = json.loads(filters)
    except:
        pass

    conditions = get_conditions(filters)


    fields = ['name', 'modified','owner','grand_total','transaction_date', 'customer', 'delivery_date',
              'expected_delivery_date', 'delivery_trip_created',
              'workflow_state', 'territory','total_qty','total_items','total_accepted_qty','department']

    query_result = frappe.get_list("Requisition", filters=conditions, fields=fields,
                                  order_by='modified desc', page_length=100)
    return query_result

def get_columns():
    columns =  [
        {'fieldname': 'check', 'label': '',"fieldtype":"Check", 'expwidth': 5, 'width': 20},
        {'fieldname': 'date', 'label': 'Date',"fieldtype":"Date", 'expwidth': 15, 'width': 100},
        {'fieldname': 'name', 'label': 'ID',"fieldtype":"Link","options" : "Requisition",'expwidth': 15, 'width': 120},
        {'fieldname': 'customer', 'label': 'Customer',"fieldtype":"Link","options" : "Customer",'expwidth': 15, 'width': 150},
        {'fieldname': 'territory', 'label': 'Territory', 'expwidth': 13, 'width': 110},
        {'fieldname': 'department', 'label': 'Department', 'expwidth': 13, 'width': 100},
        {'fieldname': 'delivery_date', 'label': 'Req Delivery',"fieldtype":"Date", 'expwidth': 13, 'width': 100, 'editable': False},
        {'fieldname': 'expected_delivery_date', 'label': 'Expt Delivery',"fieldtype":"Date", 'expwidth': 15, 'width': 100},
        {'fieldname': 'status', 'label': 'Status', 'fieldtype': 'Data', 'expwidth': 15, 'width': 140},
        {'fieldname': 'total_items', 'label': 'Items', 'fieldtype': 'Data', 'width':60},
        {'fieldname': 'total_qty', 'label': 'Reqst Qty', 'fieldtype': 'Data', 'width':80},
        {'fieldname': 'total_accepted_qty', 'label': 'Acpt Qty', 'fieldtype': 'Data', 'width':80},
        {'fieldname': 'difference', 'label': 'Diff Qty', 'fieldtype': 'Data', 'width':80},
        {'fieldname': 'delivery_trip_created', 'label': 'Has\nDT', 'fieldtype': 'Check', 'width':20},
        {'fieldname': 'action', 'label': 'View', 'fieldtype': 'Button', 'width':20, 'export': False},
        {'fieldname': 'print', 'label': 'Print', 'fieldtype': 'Button', 'width':50, 'export': False},
        {'fieldname': 'pretty_date', 'label': '', 'fieldtype': 'Data', 'width':50, 'export': False},
    ]
    return columns
def get_appropiate_action_button(requisition):
    user_roles = frappe.get_roles(frappe.session.user)
    owner = requisition.get("owner")
    action = f'''<a href="/app/requisition/{requisition.get("name")} "id="{requisition.get("name")}_Approve" target="_blank" class="btn btn-success btn-sm"
                            style="width:50px;">View</a><br>'''

    # if requisition.get("workflow_state") == "Pending for Ops Team" and "Operation" in user_roles:
    #     action = f'''<div id="{requisition.get("name")}">
    #                     <button id="{requisition.get("name")}_Approve" class="btn btn-primary btn-sm" onclick="play_action(this.id)"
    #                         style="width:66px;">Approve</button><br>
    #                     <button id="{requisition.get("name")}_Reject" class="btn btn-danger btn-sm" onclick="play_action(this.id)"
    #                         style="width:66px; margin-top:5px;">Reject</button>
    #                 </div>
    #             '''
    # elif requisition.get("workflow_state") == "Pending for Customer" and "Customer" in user_roles and frappe.session.user == owner:
    #     action = f'''<div id="{requisition.get("name")}">
    #                     <button id="{requisition.get("name")}_Approve" class="btn btn-primary btn-sm" onclick="play_action(this.id)"
    #                         style="width:66px;">Approve</button><br>
    #                     <button id="{requisition.get("name")}_Reject" class="btn btn-danger btn-sm" onclick="play_action(this.id)"
    #                         style="width:66px; margin-top:5px;">Reject</button>
    #                 </div>
    #             '''
    # elif requisition.workflow_state == "Approved" and \
    #     frappe.has_permission("Requisition", "cancel", requisition.name):

    #     action = f'''<div id="{requisition.get("name")}">
    #                     <button id="{requisition.get("name")}_Cancel" class="btn btn-danger btn-sm" onclick="play_action(this.id)"
    #                         style="width: 66px;">Cancel</button>
    #                 </div>
    #             '''
    return action

@frappe.whitelist()
def play_action(action):
    if action:
        requisition_name, action = tuple(action.split('_'))
        requisition = frappe.get_doc("Requisition", requisition_name)

        if requisition.workflow_state == 'Pending for Ops Team':
            if action == "Approve":
                requisition.workflow_state = 'Pending for Customer'
            else:
                requisition.workflow_state = 'Rejected by Ops Team'

        elif requisition.workflow_state == 'Pending for Customer':
            if action == "Approve":
                requisition.workflow_state = 'Approved'
            else:
                requisition.workflow_state = 'Rejected by Customer'


        elif requisition.workflow_state == 'Approved' and \
            frappe.has_permission("Requisition", "cancel", requisition.name) and action == 'Cancel':
            frappe.db.set_value("Requisition", requisition.name, 'workflow_state', 'Cancelled')
            frappe.db.set_value("Requisition", requisition.name, 'status', 'Cancelled')
            frappe.db.commit()

            requisition = frappe.get_doc("Requisition", requisition_name)
            requisition.cancel()

            return {
                'docname': requisition_name,
                'status': get_workflow_state_with_color(requisition),
                'action': ''
            }

        requisition.save()

        if requisition.workflow_state == "Approver" and requisition.docstatus == 0:
            requisition.submit()

        return {
            'docname': requisition_name,
            'status': get_workflow_state_with_color(requisition),
            'action': get_appropiate_action_button(requisition)
        }

def get_workflow_state_with_color(requisition):
    color = {
        "Pending for Ops Team" : "#ff5407",
        "Pending for Customer" : "#ff5557",
        "Approved" : "green",
        "Rejected by Customer" : "red",
        "Rejected by Ops Team" : "red",
        "Approved by Ops Team" : "green",
        "Cancelled" : "red"
    }
    return f'''<span style="color:{color.get(requisition.workflow_state)};">{requisition.workflow_state}</span>'''

def get_colors_dict(doctype):
    docs = frappe.get_list(doctype, fields=['name', 'color'])
    doc_colors_dict = {}

    for doc in docs:
        doc_colors_dict[doc.name] = doc.color

    return doc_colors_dict


def set_link_to_doc(doc, field, doc_url='', label=None, color=None):
    style = f'style="color:{color};"' if color else ''

    doc[field] = f'''<a href="/app/{doc_url}/{doc[field]}" target="_blank" %s>
                        {label or doc[field] or ""}
                    </a>
                ''' % style
    return doc[field]

def get_pdf_button(doctype, docname, label='', color=None):
    style = f'style="color:{color};"' if color else ''

    return f'''<a class="btn btn-sm" href="/app/print/{doctype}/{docname}" target="_blank" %s>
                        <svg class="icon  icon-sm" style="">
                            <use class="" href="#icon-printer"></use>
                        </svg>
                    </a>
            ''' % style

def custom_pretty_date(date):
    try:
        if isinstance(date, str):
            if '.' in date:
                date = date.split('.')[0]

            date = datetime.datetime.strptime(date, "%Y-%m-%d %H:%M:%S")

        pretty_date_ = pretty_date(date)

        if pretty_date_:
            pretty_date_ = pretty_date_.split(' ')

            if pretty_date_[1][0:2] == "mo":
                pretty_date_ = f"{pretty_date_[0]}M"

            pretty_date_ = f"{pretty_date_[0]}{pretty_date_[1][0]}"
            return pretty_date_
    except:
        return ""

@frappe.whitelist()
def export_file(**filters):
    columns = get_columns()
    data = get_export_data(filters)
    file_name = 'Requisition_Report.xlsx'
    generate_excel_and_download(columns, data, file_name, height=30)

def get_export_data(filters):
    query_result = get_absolute_data(filters, export=True)
    return query_result

@frappe.whitelist()
def update_field(**kwargs):
    docname = kwargs.get('docname')
    fieldname = kwargs.get('fieldname')
    value = kwargs.get('value')
    leave_application = frappe.get_doc("Merchandising Picture", docname)

    if leave_application.get(fieldname) != value:
        leave_application.update({fieldname:value})
        leave_application.save()

    return True

File: field_force/field_force/page/requisition_report/requisition_report.js
frappe.pages['requisition-report'].on_page_load = function (wrapper) {

	var page = frappe.ui.make_app_page({
	  parent: wrapper,
	  title: 'Requisition Report',
	  single_column: true
	})

	set_background_color();
	new RequisitionReport(page);

  let scrollableView = document.querySelector("[data-fieldname=scrollable_view]");
  let deliveryTripCreated = document.querySelector("[data-fieldname=delivery_trip_created]");
  scrollableView.style = 'margin-top: 30px;';
  deliveryTripCreated.style = 'margin-top: 30px;';
  }

  class RequisitionReport {
	constructor(page) {
	  this.page = page;
	  this.make_form();
	  this.make_menu();
	  // this.initialize_modal();
	  // this.open_image_modal()
	  // this.initialize_events();
	}

	initialize_events = () => {
	  this.fetch_and_render();
	}

	make_menu = () => {

	  this.page.add_action_icon("refresh", () => {
		this.fetch_and_render(true);
	  }, "Refresh");

	  this.page.add_menu_item("Export", () => {
		this.export_excel()
	  })

    if (frappe.user.has_role("Warehouse User") || frappe.user.has_role("System Manager")) {
      this.page.add_button("New Delivery Trip", () => {
        let requisition = [];
        let checkboxes = document.querySelectorAll('input[type="checkbox"]');
        checkboxes.forEach(function (checkbox) {
          if (checkbox.checked && checkbox.id !== "all-check") {
            requisition.push(checkbox.id);
          }
        });
		if (requisition.length > 0) {
			frappe.model.with_doctype("Delivery Trip", function() {
				let prospect = frappe.model.get_new_doc("Delivery Trip");
				prospect.custom_requisition_list = requisition.join(",");
				frappe.set_route("Form", "Delivery trip", prospect.name);
			});
		}else{
			msgprint("Select Requisitions First","Message")
		}
      })
    }

	  // this.page.add_action_item("PDF", () => {
		  // 	frappe.set_route('query-report', 'Employee Leave Balance Summary Report');
	  // })

	}
	make_form = () => {
	  this.form = new frappe.ui.FieldGroup({
		fields: [
			{

				fieldname: 'from_date',
				label: __('From Date'),
				fieldtype: 'Date',
				default: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
				change: () => this.fetch_and_render()
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldname: 'to_date',
				label: __('To Date'),
				fieldtype: 'Date',
				default: frappe.datetime.get_today(),
				change: () => this.fetch_and_render(),
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldname:"requisition",
				label: __("Requisition"),
				fieldtype: "Link",
				options: "Requisition",
				change: () => this.fetch_and_render(),
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldname:"customer",
				label: __("Customer"),
				fieldtype: "Link",
				options: "Customer",
				change: () => this.fetch_and_render(),
			},
			{
				fieldtype: 'Column Break'
			},
			{

				fieldname: 'territory',
				label: __('Territory'),
				fieldtype: 'Link',
				options: "Territory",
				change: () => this.fetch_and_render()
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldname:"department",
				label: __("Department"),
				fieldtype: "Link",
				options: "Customer Department",
				change: () => this.fetch_and_render(),
			},
			
			{
				fieldtype: 'Section Break'
			},
			{

				fieldname: 'status',
				label: __('Status'),
				fieldtype: 'Link',
				options: "Workflow State",
				"get_query": () => {
					return {
						filters: {
							"name": ["in" , ["Pending for Ops Team","Pending for Customer","Approved","Rejected by Customer","Rejected by Ops Team","Cancelled"]]
						}
					}
				},
				change: () => this.fetch_and_render()
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldname:"partner_group",
				label: __("Partner Group"),
				fieldtype: "Link",
				options: "Partner Group",
				change: () => this.fetch_and_render(),
			},
      {
				fieldtype: 'Column Break'
			},
      {
				fieldname: 'delivery_trip_created',
				label: __('Delivery Trip Created'),
				fieldtype: 'Check',
				default: 0,
				change: () => this.fetch_and_render(),
				hidden: false
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldname: 'scrollable_view',
				label: __('Scrollable View'),
				fieldtype: 'Check',
				default: 0,
				change: () => this.fetch_and_render(),
				hidden: false
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldtype: 'Column Break'
			},
			{
				fieldtype: 'Section Break'
			},
			{
				fieldtype: 'HTML',
				fieldname: 'preview'
			}
		],
		body: this.page.body
	  });
	  this.form.make();
	  this.fetch_and_render()
	}
	fetch_and_render = (refresh=false) => {
    let filters= this.form.get_values();
    let queryParams = window.location.search;

    if (!filters.status){
		if (queryParams) {
      const urlParams = new URLSearchParams(queryParams);
      console.log(urlParams);
      console.log(urlParams.has('workflow_state'));
      console.log(urlParams.has('delivery_trip_created'));

      if (urlParams.has('workflow_state')){
        let workflowStateValue = urlParams.get('workflow_state');
        workflowStateValue = decodeURIComponent(workflowStateValue);
        filters['status'] = workflowStateValue
        $('[data-fieldname="status"]').val(workflowStateValue)
      }

      if (urlParams.has('delivery_trip_created')){
        let deliveryTripCreated = urlParams.get('delivery_trip_created');
        deliveryTripCreated = decodeURIComponent(deliveryTripCreated);
        filters['delivery_trip_created'] = parseInt(deliveryTripCreated)
        $('[data-fieldname="delivery_trip_created"]').val(deliveryTripCreated)
      }

			window.history.replaceState({}, document.title, window.location.pathname);
		}
  }



	  // to prevent the duplicate call
	  if (refresh === false && JSON.stringify(this.filters) === JSON.stringify(filters)){
		return;
	  }
	  this.filters = this.form.get_values();

	  if (!filters.from_date) {
		this.form.get_field('preview').html('');
		return;
	  }

	  this.form.get_field('preview').html(`
			  <div class="text-muted margin-top">
				  ${__("Fetching...")}
			  </div>
		  `);
	  frappe.call('field_force.field_force.page.requisition_report.requisition_report.execute', {
		filters: filters,
		freeze: true
	  }).then(r => {
		let diff = r.message[0];
		let fields = r.message[1];
		this.render(diff, fields);
	  });
	}

	render = (diff, fields) => {
	  let table_header = this.table_header(fields);
	  let table_body = this.table_body(diff, fields);
	  let scrollable_view = this.form.get_values()['scrollable_view']

	  if (scrollable_view === 1) {
		this.form.get_field('preview').html(
		  `<div style="width:100%; overflow-x:auto; margin-top: -15px;">
			  <table class="table table-bordered" id="export_excel" style="width:160%">
					${table_header}${table_body}
			  </table>
			</div>
		  `);
	  } else {
		this.form.get_field('preview').html(
		`<div>
			  <table class="table table-bordered" id="export_excel">
					${table_header}${table_body}
			  </table>
			</div>
		`);
	  }

	}
	table_header = (headers) => {
	  let table_header = `<thead><tr>`;

	  headers.forEach(function (field, index) {
	       if (field.label == "") {
	            table_header += `<th style="padding-right: 4px;padding-left: 5px;">
                <input type="checkbox" id="all-check" onclick="toggleCheckboxes()"
                style="margin-right: 0px !important; accent-color: #00b2ff !important;"></th>`;
	       }else{
	            table_header += `<th scope="col" style="width:${field.width}px !important;">${field.label || ''}</th>`;
	       }

	  })

	  table_header += `</tr>`;
	  return table_header;
	}

	table_body = (diff, fields) => {
	  var html = `<tbody>`;
	  var serial_number = 1;

	  diff.forEach(function (data, index) {
		html += `<tr>`;

		fields.forEach(function (field, indx) {
		  if (field.fieldname === 'sl') {
			data['sl'] = index + 1;
			html += get_absolute_format_and_html(field, data);
		  }
      else if (field.editable) {
			html += `<td id="${data.docname}_${field.fieldname}" class="editable-field parent-container"
					  data-fieldname="${field.fieldname}" data-name="${data.docname}">${data[field.fieldname] || ''}</td>`
		  }
      else if (field.fieldname === 'check' && data['delivery_trip_created'] === 0 && data['workflow_state'] !== "Cancelled") {
			  html += `<td style="padding-right: 4px;padding-left: 5px;">
                    <input type="checkbox" id="${data.docname}" class="sub-checkbox" style="margin-right: 0px !important;"></td>`
		  }
      else if (field.fieldname === 'delivery_trip_created') {
        let checked = data[field.fieldname] === 1 ? "checked" : "";
			  html += `<td style="padding-right: 4px;padding-left: 5px;">
                <input type="checkbox" ${checked} id="${data.docname}" onclick="return false;"
                 style="margin-right: 0px !important; accent-color: #00b2ff !important;"></td>`
		  }
		  else {
			html += get_absolute_format_and_html(field, data);
		  }
		})

		html += `</tr>`;
		serial_number = serial_number + 1;
	  })

	  html += `</tbody>`;
	  return html
	}

	export_excel = () => {
	  let filters= this.form.get_values();
	  let url = `/api/method/field_force.field_force.page.requisition_report.requisition_report.export_file`;

	  if (filters){
		url += `?`;

		for (const field in filters) {
		  url += `${field}=${filters[field]}&`;
		}
	  }
	  window.open(url, '_blank');
	}
  }

//   amjad kalir bazar life general hospital
  function play_action(action) {
	frappe.xcall("field_force.field_force.page.requisition_report.requisition_report.play_action",
		{action: action}
	  ).then((response) => {
		let statusFieldID = response.docname + '_status';
		document.getElementById(response.docname).innerHTML = response.action;
		document.getElementById(statusFieldID).innerHTML = response.status;
	  });
  }

  function get_absolute_format_and_html(field, data, width = 50) {
	let docname = data.docname;
	let value = data[field.fieldname];

	if (field.fieldtype === "Image") {
	  return get_image_html(value);
	} else if (field.fieldtype === "Currency") {
	  return get_currency_format(value);
	} else if (field.fieldtype === "Data") {
	  return `<td id="${docname}_${field.fieldname}">${value || ''}</td>`;
	} else {
	  return `<td id="${docname}_${field.fieldname}">${value || ''}</td>`;
	}
  }

  function get_image_html(image_url) {
	return `
		  <td style="height:100px; width:120px;">
			  <a href="#">
				  <img style="height:100%; width:100%" id="image" src="${image_url}" onclick="(
					  function(e){
						  document.getElementById(\'modal_section\').style.display=\'block\';
						  var nAgt = navigator.userAgent;
						  if (nAgt.indexOf('Safari') !== -1) {
							  document.getElementById(\'img\').src=e.target.currentSrc;
						  }else{
							  document.getElementById(\'img\').src=e.path[0].currentSrc;
						  }
						  return false;
					  }
				  )(arguments[0]);
				  return false;
			  </a>
		  </td>`;
  }

  function get_currency_format(value) {
	const company = frappe.defaults.get_user_default("company");
	const currency = frappe.get_doc(":Company", company).default_currency;
	let value_in_currency = format_currency(value, currency);
	return `<td>${value_in_currency || ''}</td>`
  }

  function set_background_color(){
	let current_theme = document.documentElement.getAttribute("data-theme")
	if (current_theme === 'light'){
		$('.page-body').css('background', '#FFFFFF');
	} else{
		$('.page-body').css('background', 'black');
	}
  }
  function dateFormat(input_D, format_D) {
    const date = new Date(input_D);
    const day = date.getDate();
    const month = date.getMonth() + 1;
    const year = date.getFullYear();
    format_D = format_D.replace("MM", month.toString().padStart(2,"0"));
    if (format_D.indexOf("yyyy") > -1) {
        format_D = format_D.replace("yyyy", year.toString());
    } else if (format_D.indexOf("yy") > -1) {
        format_D = format_D.replace("yy", year.toString().substr(2,2));
    }
    format_D = format_D.replace("dd", day.toString().padStart(2,"0"));
    return format_D;
}
function toggleCheckboxes() {
    var mainCheckbox = document.getElementById("all-check");
    var checkboxes = document.getElementsByClassName("sub-checkbox");

    // If main checkbox is checked, check all sub-checkboxes
    if (mainCheckbox.checked) {
        for (var i = 0; i < checkboxes.length; i++) {
            checkboxes[i].checked = true;
        }
    } else {
        // If main checkbox is unchecked, uncheck all sub-checkboxes
        for (var i = 0; i < checkboxes.length; i++) {
            checkboxes[i].checked = false;
        }
    }
}
